src/lxc/lxc_user_nic.c

```diff
@@ -96,20 +97,157 @@ static char *get_username(void)
 	return pwd->pw_name;
 }
 
+static char **get_groupnames(void)
+{
+	int ngroups;
+	gid_t *group_ids;
+	int ret, i, j;
+	char **groupnames;
+	struct group *gr;
+
+	ngroups = getgroups(0, NULL);
+
+	if (ngroups == -1) {
+		fprintf(stderr, "Failed to get number of groups user belongs to\n");
+		return NULL;
+	}
+
+	group_ids = (gid_t *)malloc(sizeof(gid_t)*ngroups);
+	ret = getgroups(ngroups, group_ids);
+
+	if (ret < 0) {
+		free(group_ids);
+		fprintf(stderr, "Failed to get process groups\n");
+		return NULL;
+	}
+
+	groupnames = (char **)malloc(sizeof(char *)*(ngroups+1));
+
+	for (i=0; i<ngroups; i++ ) {
+		gr = getgrgid(group_ids[i]);
+
+		if (gr == NULL) {
+			fprintf(stderr, "Failed to get group name\n");
+			free(group_ids);
+			for (j=0; j<i; j++) {
+				free(groupnames[j]);
+			}
+			free(groupnames);
```

Comment:
I think that it is safer not to, since free_groupnames() relies on the list being null-terminated which it isn't yet at that point.